src/materials/MazarsDamage.C

```diff
@@ -0,0 +1,145 @@
+    _kappa_old(getMaterialPropertyOld<Real>("kappa")),
+    _damage_index(declareProperty<Real>(_base_name + "damage_index")),
+    _damage_index_old(getMaterialPropertyOld<Real>("damage_index")),
+    _stress(getMaterialProperty<RankTwoTensor>(_base_name + "stress")),
+    _mechanical_strain(getMaterialProperty<RankTwoTensor>(_base_name + "mechanical_strain")),
+    _elasticity_tensor_name(_base_name + "elasticity_tensor"),
+    _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_elasticity_tensor_name)),
+    _eigval(3, 0.0),
+    _positive_stress(3, 0.0),
+    _positive_strain(3, 0.0)
+{
+}
+
+void
+MazarsDamage::initQpStatefulProperties()
+{
+  _damage_index[_qp] = 0.0;
+  _kappa[_qp] = 0.0;
+}
+
+void
+MazarsDamage::initialSetup()
+{
+  if (!hasGuaranteedMaterialProperty(_elasticity_tensor_name, Guarantee::ISOTROPIC))
+    mooseError("MazarsDamage requires that the elasticity tensor be guaranteed isotropic");
+}
+
+void
+MazarsDamage::updateDamage()
+{
+  _stress[_qp].symmetricEigenvalues(_eigval);
+  for (unsigned int i = 0; i < 3; ++i)
+    _positive_stress[i] = std::max(_eigval[i], 0.0);
+
+  _mechanical_strain[_qp].symmetricEigenvalues(_eigval);
+  for (unsigned int i = 0; i < 3; ++i)
+    _positive_strain[i] = std::max(_eigval[i], 0.0);
+
+  const Real equiv_tensile_strain =
+      std::sqrt(Utility::pow<2>(_positive_strain[0]) + Utility::pow<2>(_positive_strain[1]) +
+                Utility::pow<2>(_positive_strain[2]));
+
+  Real alpha_t = 0.0;
+  Real alpha_c = 0.0;
+
+  const Real e = ElasticityTensorTools::getIsotropicYoungsModulus(_elasticity_tensor[_qp]);
+  const Real nu = ElasticityTensorTools::getIsotropicPoissonsRatio(_elasticity_tensor[_qp]);
+
+  const Real sum_pos_stress = _positive_stress[0] + _positive_stress[1] + _positive_stress[2];
+
+  for (unsigned int i = 0; i < 3; ++i)
+  {
+    const Real epsilon_ti = ((1.0 + nu) * _positive_stress[i] - nu * sum_pos_stress) / e;
+    const Real epsilon_ci = _eigval[i] - epsilon_ti; //_eigval contains the principal strains
+    if (!MooseUtils::absoluteFuzzyEqual(equiv_tensile_strain, 0.0))
+    {
+      const Real ets2 = Utility::pow<2>(equiv_tensile_strain);
+      alpha_t += epsilon_ti * _positive_strain[i] / ets2;
+      alpha_c += epsilon_ci * _positive_strain[i] / ets2;
+    }
+  }
+
+  const Real kappa_0 = _tensile_strength / e; // Threshold strain for damage
+  const Real kappa = std::max(kappa_0, _kappa_old[_qp]);
+
+  _damage_index[_qp] = 0.0;
+  if (equiv_tensile_strain > kappa)
+  {
+    _kappa[_qp] = equiv_tensile_strain;
+    const Real kappa = _kappa[_qp];
+    const Real tensile_damage =
+        1.0 - kappa_0 * (1.0 - _a_t) / kappa - _a_t * std::exp(-_b_t * (kappa - kappa_0));
+    const Real compressive_damage =
+        1.0 - kappa_0 * (1.0 - _a_c) / kappa - _a_c * std::exp(-_b_c * (kappa - kappa_0));
+    _damage_index[_qp] = std::max(alpha_t * tensile_damage + alpha_c * compressive_damage, 0.0);
+  }
```

Comment:
It seems that we might end in a situation where the new ``damage_index`` becomes lower than its old value (for example, if we unload the specimen, or if ``e`` (and therefore ``kappa_0``) changes for whatever reason. We probably need an additional check ``_damage_index[_qp] = std::max(_damage_index[_qp], _damage_index_old[_qp]`` to prevent that from happening.